engine/language_client_go/main.go

```diff
@@ -0,0 +1,109 @@
+package main
+
+/*
+#cgo LDFLAGS: ./lib/libhello.dylib -ldl
+#include <stdlib.h>
+#include <stdbool.h>
+#include "./lib/baml.h"
+
+extern void trigger_callback(uint32_t, bool, const char *);
+*/
+import "C"
+
+import (
+	"fmt"
+	"unsafe"
+)
+
+type StreamResult[Partial any, Final any] struct {
+	partial *Partial
+	final   *Final
+	error   error
+}
+
+func (result *StreamResult[Partial, Final]) Partial() Partial {
+	return *result.partial
+}
+
+func (result *StreamResult[Partial, Final]) Final() Final {
+	return *result.final
+}
+
+func (result *StreamResult[Partial, Final]) IsFinal() bool {
+	return result.final != nil
+}
+
+func (result *StreamResult[Partial, Final]) IsPartial() bool {
+	return result.partial != nil
+}
+
+func (result *StreamResult[Partial, Final]) Error() error {
+	return result.error
+}
+
+type DoSomethingStreamResult struct {
+	StreamResult[string, string]
+}
+
+// UpperCase in Go only!
+func DoSomethingStream(arg string) <-chan DoSomethingStreamResult {
```

Comment:
I'd be cool if we returned StreamResult directly here. Finalize the return types in the declaration here itself.